### PR TITLE
Esi 2220 hooks support

### PR DESCRIPTION
**Pull request by RamJogal**

### CODE REVIEW

File: ui/src/containers/CustomField/Product.tsx [hunk 1]
```diff
@@ -83,57 +82,80 @@ const Product: React.FC<Props> = function ({ product, remove }) {
         }),
       className: "ActionListItem--warning",
     },
-  ];
+  ]?.filter(
+    (action) => !(isProductDeleted && action?.label?.props?.icon === "NewTab")
```
Comment (NamrataSane) [inline ui/src/containers/CustomField/Product.tsx:86]: "NewTab" --> add in the locale

File: ui/src/containers/CustomField/RenderList.tsx [hunk 2]
```diff
@@ -4,14 +4,10 @@ import DraggableGrid from "./DraggableGrid";
...
-  type,
-  // selectedItems = [],
-  childWindow,
-  view,
-}) {
+const RenderList: React.FC<Props> = function ({ type, childWindow, view }) {
   const { selectedItems, removeIdFromField } = useProductCustomField();
-  const removeItem = (removeId: any) => {
+  const removeItem = (removeId: any, multiConfigName: any) => {
+    console.info("multiConfigName", multiConfigName);
```
Comment (NamrataSane) [inline ui/src/containers/CustomField/RenderList.tsx:10]: remove console

File: ui/src/containers/CustomField/styles.scss [hunk 3]
```diff
@@ -112,7 +112,7 @@
         display: grid;
         grid-template-columns: 1fr 1fr;
         overflow-y: auto;
-        overflow-x: hidden;
+        // overflow-x: hidden;
```
Comment (NamrataSane) [inline ui/src/containers/CustomField/styles.scss:115]: remove commented code

File: ui/src/containers/SelectorPage/index.tsx [hunk 4]
```diff
@@ -64,6 +65,46 @@ const SelectorPage: React.FC = function () {
...
+            const multiConfigKeys: any = [];
+            data?.advancedConfig?.forEach((keys: any) => {
+              const obj = {
+                label: keys,
+                value: keys,
+                default: keys === data?.advancedConfig?.[0],
+              };
+              multiConfigKeys.push(obj);
+              setMultiConfigDropDown([...multiConfigKeys]);
+              if (multiConfigKeys.length) {
```
Comment (NamrataSane) [inline ui/src/containers/SelectorPage/index.tsx:81]: multiConfigKeys?.

File: ui/src/containers/SelectorPage/index.tsx [hunk 5]
```diff
@@ -64,6 +65,46 @@ const SelectorPage: React.FC = function () {
...
+                multiConfigKeys[0].default = true;
+                setSelectedMultiConfigValue(multiConfigKeys[0]);
```
Comment (NamrataSane) [inline ui/src/containers/SelectorPage/index.tsx:83]: multiConfigKeys?.[0]

File: ui/src/root_config/categories/index.ts [hunk 6]
```diff
@@ -3,17 +3,24 @@ import rootConfig from "..";
...
-    fieldData?.data?.map((i: any) => ({
-      [rootConfig.ecommerceEnv.UNIQUE_KEY.category]:
-        i?.[rootConfig.ecommerceEnv.UNIQUE_KEY.category],
-      catalogId: i?.catalogId,
-      catalogVersionId: i?.catalogVersionId,
-    })),
-  fetchCustomCategoryData: (config: any, type: any, selectedIDs: any) => {
+    fieldData?.data?.map(
+      (i: any) =>
+        // [rootConfig.ecommerceEnv.UNIQUE_KEY.category]:
```
Comment (NamrataSane) [inline ui/src/root_config/categories/index.ts:8]: remove commented code

File: ui/src/root_config/categories/index.ts [hunk 7]
```diff
@@ -3,17 +3,24 @@ import rootConfig from "..";
...
+        i?.[rootConfig.ecommerceEnv.UNIQUE_KEY.category]
+      // catalogId: i?.catalogId,
```
Comment (NamrataSane) [inline ui/src/root_config/categories/index.ts:10]: remove commented code

File: ui/src/root_config/categories/index.ts [hunk 8]
```diff
@@ -3,17 +3,24 @@ import rootConfig from "..";
...
+      // catalogVersionId: i?.catalogVersionId,
```
Comment (NamrataSane) [inline ui/src/root_config/categories/index.ts:11]: remove commented code

File: ui/src/root_config/index.ts [hunk 9]
```diff
@@ -552,19 +569,70 @@ const removeItemsFromCustomField = (
...
+    if (type === "category")
+      return selectedIds?.filter((data: any) => data?.[uniqueKey] !== removeId);
+
+    return selectedIds?.filter(
+      (data: any) => Number(data) !== Number(removeId)
+    );
+  }
+  let updatedRootConfig = { ...selectedIds };
+  const config = selectedIds?.[multiConfigName];
+  const updatedIds = config.multiConfiguniqueKey.filter(
```
Comment (NamrataSane) [inline ui/src/root_config/index.ts:586]: ?.filter

File: ui/src/root_config/index.ts [hunk 10]
```diff
@@ -552,19 +569,70 @@ const removeItemsFromCustomField = (
...
+  updatedRootConfig = {
+    ...updatedRootConfig,
+    [multiConfigName]: updatedConfig,
+  };
+  return updatedRootConfig;
+};
 
-  return selectedIds?.filter((data: any) => Number(data) !== Number(removeId));
+const returnFormattedProductIDSForMultiConfig = (productData: any): Result => {
+  console.info("productData", productData);
```
Comment (NamrataSane) [inline ui/src/root_config/index.ts:602]: remove console

File: ui/src/root_config/index.ts [hunk 11]
```diff
@@ -552,19 +569,70 @@ const removeItemsFromCustomField = (
...
+  let result: Result = {};
+  if (productData?.length) {
+    const uniqueKey = "multiConfiguniqueKey";
+    result = productData.reduce((acc: any, item: any) => {
```
Comment (NamrataSane) [inline ui/src/root_config/index.ts:606]: productData?.

File: ui/src/root_config/index.ts [hunk 12]
```diff
@@ -552,19 +569,70 @@ const removeItemsFromCustomField = (
...
+      const multiConfigName = item?.cs_metadata?.multi_config_name;
+      // console.info("multiConfigName",multiConfigName)
```
Comment (NamrataSane) [inline ui/src/root_config/index.ts:608]: remove commented code

File: ui/src/root_config/index.ts [hunk 13]
```diff
@@ -552,19 +569,70 @@ const removeItemsFromCustomField = (
...
+      if (!acc[multiConfigName]) {
+        acc[multiConfigName] = { [uniqueKey]: [] };
+      }
+      acc[multiConfigName][uniqueKey].push(item.id);
```
Comment (NamrataSane) [inline ui/src/root_config/index.ts:612]: acc?.[multiConfigName]?.[uniqueKey]?..push(item?.id);

File: ui/src/root_config/index.ts [hunk 14]
```diff
@@ -552,19 +569,70 @@ const removeItemsFromCustomField = (
...
+  return result;
+};
+
+const returnFormattedCategoryIDSForMultiConfig = (
+  categoryData: any
+): Result => {
+  let result: Result = {};
+  if (categoryData?.length) {
+    const uniqueKey = "multiConfiguniqueKey";
+    result = categoryData.reduce((acc: any, item: any) => {
```
Comment (NamrataSane) [inline ui/src/root_config/index.ts:625]: categoryData?.

File: ui/src/root_config/index.ts [hunk 15]
```diff
@@ -552,19 +569,70 @@ const removeItemsFromCustomField = (
...
+      const multiConfigName = item?.cs_metadata?.multi_config_name;
+      console.info("");
```
Comment (NamrataSane) [inline ui/src/root_config/index.ts:627]: remove console

File: ui/src/root_config/index.ts [hunk 16]
```diff
@@ -552,19 +569,70 @@ const removeItemsFromCustomField = (
...
+      if (!acc[multiConfigName]) {
+        acc[multiConfigName] = { [uniqueKey]: [] };
+      }
+      acc[multiConfigName][uniqueKey].push(item.id);
```
Comment (NamrataSane) [inline ui/src/root_config/index.ts:631]: acc?.[multiConfigName]?.[uniqueKey]?.push(item?.id);

File: ui/src/root_config/index.ts [hunk 17]
```diff
@@ -580,7 +648,10 @@ const rootConfig = {
   categorySelectorColumns,
   getCustomKeys,
   getSidebarData,
+  // arrangeList,
```
Comment (NamrataSane) [inline ui/src/root_config/index.ts:651]: remove commented code

File: ui/src/services/index.ts [hunk 18]
```diff
@@ -56,28 +62,34 @@ const requestCategories = (config: any) =>
...
-        config
-      )
-    : null;
+const getSelectedIDs = async (
+  config: any,
+  type: any,
+  selectedIDs: any,
+  isOldUser: any
+) => {
+  const ids =    isOldUser === true ? selectedIDs.join(",") : JSON.stringify(selectedIDs);
```
Comment (NamrataSane) [inline ui/src/services/index.ts:71]: selectedIDs?.join

File: ui/src/services/index.ts [hunk 19]
```diff
@@ -56,28 +62,34 @@ const requestCategories = (config: any) =>
...
   config: any,
   type: any,
-  selectedIDs: any
+  selectedIDs: any,
+  isOldUser: any
 ) => {
-  if (Array.isArray(selectedIDs) && selectedIDs.length) {
+  if (
+    isOldUser
+      ? Array.isArray(selectedIDs) && selectedIDs.length
```
Comment (NamrataSane) [inline ui/src/services/index.ts:85]: selectedIDs?.

File: ui/src/containers/ConfigScreen/index.tsx [hunk 20]
```diff
@@ -57,102 +75,381 @@ const ConfigScreen: React.FC = function () {
...
+  const checkValidity = async (
+    configurationData: { [x: string]: any; multi_config_keys?: any },
+    serverConfiguration: { [x: string]: any; multi_config_keys?: any },
+    validateMultiConfigKeysByApi: boolean,
+    validateOtherKeysByApi: boolean
+  ) => {
+    const checkMultiConfigKeys = (multiConfigKeys: any) => {
+      const invalidKeys: any = {};
+
+      Object.entries(multiConfigKeys || {}).forEach(([configKey, config]) => {
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:163]: ?.forEach

File: ui/src/containers/ConfigScreen/index.tsx [hunk 21]
```diff
@@ -57,102 +75,381 @@ const ConfigScreen: React.FC = function () {
...
+        .map(([source, keys]) => ({ source, keys })),
+    ];
+
+    let otherInvalidKeysList: { source: string; keys: any[] }[] = [];
+    if (!validateOtherKeysByApi) {
+      const otherInvalidKeys = checkOtherKeys(configurationData);
+      const serverOtherInvalidKeys = checkOtherKeys(serverConfiguration);
+
+      otherInvalidKeysList = [
+        ...otherInvalidKeys.map((key) => ({
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:216]: otherInvalidKeys?.

File: ui/src/containers/ConfigScreen/index.tsx [hunk 22]
```diff
@@ -57,102 +75,381 @@ const ConfigScreen: React.FC = function () {
...
+          source: "configurationData",
+          keys: [key],
+        })),
+        ...serverOtherInvalidKeys.map((key) => ({
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:220]: serverOtherInvalidKeys?.

File: ui/src/containers/ConfigScreen/index.tsx [hunk 23]
```diff
@@ -57,102 +75,381 @@ const ConfigScreen: React.FC = function () {
...
+
+      const apiInvalidKeysList = apiValidationResults.invalidKeys;
+
+      const allInvalidKeys = [
+        ...normalInvalidKeysList,
+        ...otherInvalidKeysList,
+        ...apiInvalidKeysList,
+      ];
+
+      const isValid = allInvalidKeys.length === 0;
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:247]: allInvalidKeys?.

File: ui/src/containers/ConfigScreen/index.tsx [hunk 24]
```diff
@@ -57,102 +75,381 @@ const ConfigScreen: React.FC = function () {
...
+                        ?.legacy_config?.[key] !== undefined
+                        ? installationDataOfSdk?.configuration
+                            ?.multi_config_keys?.legacy_config?.[key]
+                        : installationDataOfSdk?.configuration?.[key];
+                  }
+                }
+              }
+            );
+
+            Object.keys(installationDataOfSdk?.serverConfiguration).forEach(
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:369]: ?.forEach

File: ui/src/containers/ConfigScreen/index.tsx [hunk 25]
```diff
@@ -57,102 +75,381 @@ const ConfigScreen: React.FC = function () {
...
+                  acc[key] = installationDataOfSdk?.configuration?.[key];
+                }
+                return acc;
+              }, {});
+
+            const filteredServerConfiguration: any =              combinedServerConfigurationKeys?.reduce((acc: any, key: any) => {
+                if (
+                  Object.hasOwn(installationDataOfSdk?.serverConfiguration, key)
+                ) {
+                  acc[key] = installationDataOfSdk?.serverConfiguration[key];
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:399]: installationDataOfSdk?.serverConfiguration?.[key];

File: ui/src/containers/ConfigScreen/index.tsx [hunk 26]
```diff
@@ -226,62 +566,486 @@ const ConfigScreen: React.FC = function () {
...
+
+    setState((prevState: any) => {
+      updatedConfigAccordions = {
+        ...prevState?.installationData?.configuration?.multi_config_keys,
+      };
+      updatedServerConfigAccordions = {
+        ...prevState?.installationData?.serverConfiguration?.multi_config_keys,
+      };
+
+      delete updatedConfigAccordions[id];
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:641]: delete updatedConfigAccordions?.[id];

File: ui/src/containers/ConfigScreen/index.tsx [hunk 27]
```diff
@@ -226,62 +566,486 @@ const ConfigScreen: React.FC = function () {
...
+      delete updatedServerConfigAccordions[id];
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:642]: delete updatedServerConfigAccordions?.[id];

File: ui/src/containers/ConfigScreen/index.tsx [hunk 28]
```diff
@@ -226,62 +566,486 @@ const ConfigScreen: React.FC = function () {
...
+                title={localeTexts.configPage.multiConfig.accordionLabel?.replace(
+                  "$",
+                  rootConfig.ecommerceEnv.APP_ENG_NAME
+                )}
+                renderExpanded
+                accordionDataCount={
+                  Object.keys(
+                    state?.installationData?.configuration?.multi_config_keys
+                      || {}
+                  ).length
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:741]: )?.length

File: ui/src/containers/ConfigScreen/index.tsx [hunk 29]
```diff
@@ -226,62 +566,486 @@ const ConfigScreen: React.FC = function () {
...
+                                        version="v2"
+                                      />
+                                    </Dropdown>
+                                  </div>
+                                ),
+                                onClick: () => {},
+                              },
+                            ]}
+                          >
+                            {Object.entries(configInputFields).map(
```
Comment (NamrataSane) [inline ui/src/containers/ConfigScreen/index.tsx:857]: ?.map(

File: ui/src/common/providers/ProductCustomFieldExtensionProvider.tsx [hunk 30]
```diff
@@ -0,0 +1,286 @@
...
+  );
+
+  const fetchData = async (selectedIdsArray: any) => {
+    if (
+      isOldUser
+        ? Array.isArray(selectedIdsArray)
+        : Object.keys(selectedIdsArray)?.length
+          && !isEmpty(appConfig)
+          && isOldUser
+        ? selectedIdsArray.length
```
Comment (NamrataSane) [inline ui/src/common/providers/ProductCustomFieldExtensionProvider.tsx:195]: selectedIdsArray?.

File: api/root_config/index.ts [hunk 31]
```diff
@@ -20,75 +19,205 @@ const root_config: any = {
...
+          return categoryResponse;
+        })
+      );
+    }
   },
 
   getSelectedProductsandCategories: (data: any, key: any) => {
     let url: string = root_config.getUrl(key, data?.query);
-    const urlHasQueryParams: boolean = url.indexOf("?") > -1;
+    const urlHasQueryParams: boolean = url.includes("?");
```
Comment (NamrataSane) [inline api/root_config/index.ts:131]: url?.

**Review by NamrataSane — CHANGES_REQUESTED** [review]
Resolve given comments

File: ui/src/containers/ConfigScreen/index.tsx [hunk 32]
```diff
@@ -57,102 +75,381 @@ const ConfigScreen: React.FC = function () {
...
+      normalInvalidKeys = checkMultiConfigKeys(
+        configurationData?.multi_config_keys
+      );
+      serverNormalInvalidKeys = checkMultiConfigKeys(
+        serverConfiguration?.multi_config_keys
+      );
+    }
+
+    const normalInvalidKeysList = [
+      ...Object.entries(normalInvalidKeys)
```
[the remark this answers is not in the document]
Comment (RamJogal) [reply ui/src/containers/ConfigScreen/index.tsx:202]: After adding optional chaining getting an  Unsafe usage of optional chaining. If it short-circuits with 'undefined' the evaluation will throw TypeError

File: ui/src/containers/ConfigScreen/index.tsx (hunk 21 above)
Comment (RamJogal) [reply ui/src/containers/ConfigScreen/index.tsx:216]: After adding optional chaining getting an Unsafe usage of optional chaining. If it short-circuits with 'undefined' the evaluation will throw TypeError

File: ui/src/containers/ConfigScreen/index.tsx (hunk 22 above)
Comment (RamJogal) [reply ui/src/containers/ConfigScreen/index.tsx:220]: After adding optional chaining getting an Unsafe usage of optional chaining. If it short-circuits with 'undefined' the evaluation will throw TypeError

**Review by NamrataSane — APPROVED** [review]

**Pull request merged**